mongo/options/findoptions.go

```diff
@@ -242,6 +258,12 @@ func (f *FindOptions) SetSort(sort interface{}) *FindOptions {
 	return f
 }
 
+// SetTimeout sets the value for the Timeout field.
```

Comment:
I argue to omit `SetTimeout` on operations. Users can achieve this with a context and using `SetTimeout(0)` on the Collection / Database / Client to opt into the backwards breaking behavior:
```go
ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
defer cancel()
coll := db.Collection("coll", options.Collection().SetTimeout(0))
cursor, err := coll.Find(ctx, bson.D{})
```

In a 2.0 release, the backwards breaking behavior gated by the `Timeout` option may become the default. The operation `Timeout` would not enable anything more than a context with a timeout. 

An operation Timeout option is required in the current specification. @ShaneHarvey do you think it is OK to pursue a change in CSOT like the following?

> This value MUST be configurable at the level of a MongoClient, MongoDatabase, MongoCollection, or of a single operation.

To:

> This value MUST be configurable at the level of a MongoClient, MongoDatabase, MongoCollection, or of a single operation. Drivers for languages that enable an existing operation timeout (e.g. a `context` in Go) MAY use the existing operation timeout in place of a new `timeoutMS` option.

Reply:
That sounds good to me. I've removed the `Timeout` field from `FindOptions` and the associated setter.

Reply:
I like the idea of using the idiomatic timeout approach rather than introducing a new way to do it but I'll have to check with others. Would Go still support the timeoutMS URI option? 

Can you also explain the "backwards breaking behavior"? Why would `options.Collection().SetTimeout(0)` be required?

Reply:
> Would Go still support the timeoutMS URI option?

Yep, the logic is already present in [connstring.go](https://github.com/mongodb/mongo-go-driver/pull/863/files/b6726b4f0866b64fe987881476ae2270e3346724#diff-33b05fea5109bc786233b06b530c7f78a4f0deb9b13f12770f296d61c8f767a6R888-R894).

> Can you also explain the "backwards breaking behavior"?

Setting `Timeout` in the Go driver at any level will ignore the values of certain, deprecated options such as `SocketTimeout`, `wTimeout` and `MaxTimeMS`. This is hypothetically backwards-breaking, and in version 2.0 of the driver, these deprecated options will be removed and only `Timeout` will remain.

> Why would `options.Collection().SetTimeout(0)` be required?

A `Timeout` of 0 is technically an infinite timeout, and would be the same as having no set timeout in the Go driver. Explicitly setting 0, though, would opt users into the backwards-breaking behavior of ignoring the deprecated options, so it could be a useful mechanic.

Reply:
I assume you mean separating the "opting-in" to backwards-breaking behavior and setting the timeout with two different options? An interesting thought, what are your plans in the Python driver?

Reply:
We don't have the same problem because Python doesn't have a Context timeout mechanism.

Reply:
I'm not sure "opting-in" makes sense in the Go driver API. Consider the following client configuration:
```go
options.Client().SetTimeout(1 * time.Minute).SetSocketTimeout(30 * time.Second).UseNewTimeoutBehavior(true)
```

That is basically configuring a client to enable a global timeout and a socket timeout then immediately ignoring the socket timeout. Separately "opting-in" to the new behavior would really just mean removing the now-deprecated timeout configurations:
```go
options.Client().SetTimeout(1 * time.Minute)
```

Reply:
I agree: I don't love the separate opt-in, as it introduces an extra knob and (I think) increases code complexity. And, what would we do in the following case?
```
options.Client().SetTimeout(1 * time.Minute).SetSocketTimeout(30 * time.Second).UseNewTimeoutBehavior(false)
```
I _think_ we'd want to honor the deprecated option since it's "further down", and the user is explicitly not using the new timeout behavior? That's extra logic for all of the deprecated options, and again, we're just setting an option (`SetTimeout`) that is immediately ignored. As you said, this behavior requires more logic in the driver, introduces confusing client configurations and represents another knob.

Reply:
We could make setting a global timeout and deprecated timeout options together a configuration error. There's not a good way to enforce that configuration error when setting a timeout at the Database or Collection level, so we would have to remove those settings and only set timeout on a Client or an individual operation.

Is there a lot of value to being able to set a timeout at the Database or Collection level? Is it worth the added confusion of allowing contradictory configurations?

Reply:
Interesting idea. The only conflicting timeout option at the `Client` level is `SocketTimeout`; I suppose we could throw a configuration error there (add some logic in `ClientOptions.validateAndSetError()`). The other deprecated timeout options: `wTimeout`, `MaxTime` and `MaxCommitTime` are not `ClientOptions` and couldn't be recognized as configuration errors in the same way, I think.

> Is there a lot of value to being able to set a timeout at the Database or Collection level?

I can imagine a use-case where a user wants a particular `Timeout` on a `Database` or `Collection` instead of the entire `Client`. We'd be very much against spec if we _only_ allowed specifying `Timeout` on `Client`.

Reply:
Ah, I was actually proposing we change the spec to disallow setting timeout at the Database and Collection level, which would allow us to do more validation at Client initialization. However, considering we would only be able to clearly validate that `Timeout` and `SocketTimeout` are mutually exclusive (and not `wTimeout`, `MaxTime`, and `MaxCommitTime`), I don't think it's worth the reduction in functionality.